react_juce/core/EcmascriptEngine_Hermes.cpp

```diff
@@ -0,0 +1,522 @@
+#include "EcmascriptEngine.h"
+
+#include <hermes/hermes.h>
+
+using namespace facebook;
+
+namespace reactjuce
+{
+    namespace
+    {
+        //==============================================================================
+        class JSIMemoryBuffer : public jsi::Buffer
+        {
+        public:
+            explicit JSIMemoryBuffer(std::unique_ptr<juce::MemoryBlock> mem)
+                : memBlock(std::move(mem))
+            { }
+
+            size_t size() const override
+            {
+                return memBlock->getSize();
+            }
+
+            const uint8_t *data() const override
+            {
+                return reinterpret_cast<const uint8_t*>(memBlock->getData());
+            }
+
+        private:
+            std::unique_ptr<juce::MemoryBlock> memBlock;
+        };
+
+        //==============================================================================
+        struct CopyableJSIMethodWrapper
+        {
+            explicit CopyableJSIMethodWrapper(jsi::Function f)
+                : fn(std::move(f))
+            { }
+
+            jsi::Function fn;
+        };
+
+        //==============================================================================
+        juce::var  jsiValueToVar(const jsi::Value &v, jsi::Runtime &runtime);
+        jsi::Value varToJSIValue(const juce::var &v, jsi::Runtime &runtime);
+
+        juce::var jsiArrayToVarArray(const jsi::Array &v, jsi::Runtime &runtime)
+        {
+            juce::Array<juce::var> varArray;
+
+            const size_t numItems = v.size(runtime);
+            for (size_t i = 0; i < numItems; ++i)
+            {
+                varArray.add(jsiValueToVar(v.getValueAtIndex(runtime, i), runtime));
+            }
+
+            return varArray;
+        }
+
+        juce::var::NativeFunction jsiMethodToVarMethod(jsi::Function v, jsi::Runtime &runtime)
+        {
+            auto fPtr = std::make_shared<CopyableJSIMethodWrapper>(std::move(v));
+
+            return [fPtr = fPtr, &rt = runtime] (const juce::var::NativeFunctionArgs &args)
```

Comment:
Do we only need the `CopyableJSIMethodWrapper` helper so that we can capture here? If so, I think we can clean it up by moving in the capture group:

```
return [fPtr = std::move(v), &rt = runtime](...blah...)
```

If that works, let's go for it and drop the CopyableJSIMethodWrapper code

Reply:
It's actually a result of returning the lambda and associated capture block from this function. This is the only way I could get around it after much stack overlow. Tried moving the jsi::Function and didn't work. See here: https://stackoverflow.com/questions/24375342/returning-a-lambda-function-with-a-move-capture

This is a result of `juce::var::NativeFunction` being typed as `std::function` object.